Added functionality for pulling audit logs

fishygeek91 opened the pull request:
Created new fields "Last Success" and "First Failure" on custom object "Test_Run_Method_Results__c". Also added formula field to this custom object that generates audit query.

Added method processPassFailDates() and used it within processTestResults()

fishygeek91 committed:
-Added DateTime for first failure/ last success and field for failure audit

ChuckJonas commented on force-app/main/default/classes/TestRunProcessor.cls:
```diff
@@ -103,11 +103,40 @@ public with sharing class TestRunProcessor implements Schedulable {
+        for (Test_Run_Method_Result__c r : result) {
+            if (!r.Method_Pass__c) {
+                r.First_Failure__c = r.CreatedDate;
+                for (Test_Run_Method_Result__c f : failedCousins) {
```
Instead of the nested for each, would be much more efficient to create a `Map<String, List<Test_Run_Method_Result__c>` where the key is the `Method_Name__c` ahead of time.  The you can just loop through each `result` and get all the matching test methods by calling `pastResults.get(r.Method_Name__c)`.

You'll still end up with nested loops but the second loop will be on a MUCH smaller collection.

fishygeek91 replied:
Ok. I added a map instead. Came up with the same results so that's good. Had to add another for loop to create the map. Couldn't come up with a more efficient way, but maybe you have one in mind

ChuckJonas replied:
nope, this is the most efficient.  With this code you only see each item at most twice (once to create the map and when processing the failures).  Your previous implementation would visit each item on every failure (worst case).

ChuckJonas commented on force-app/main/default/classes/TestRunProcessor.cls:
```diff
@@ -103,11 +103,40 @@ public with sharing class TestRunProcessor implements Schedulable {
-
-        insert results;
+        
+		List<Test_Run_Method_Result__c> result = processPassFailDates(results);
```
looks like you've got some whitespace issues (tabs vs spaces).  Did you run `npm install` to get the prettier cli?

fishygeek91 replied:
fixed

ChuckJonas commented on force-app/main/default/classes/TestRunProcessor.cls:
```diff
@@ -103,11 +103,40 @@ public with sharing class TestRunProcessor implements Schedulable {
+        List<Test_Run_Method_Result__c> failedCousins = [SELECT Id, Method_Name__c, Method_Pass__c, CreatedDate FROM Test_Run_Method_Result__c 
+                                                   WHERE Method_Name__c IN :failedTests ORDER BY CreatedDate DESC];
+        
+        for (Test_Run_Method_Result__c r : result) {
```
shouldn't we just be iterating over `failedTests` here?

fishygeek91 replied:
No because I am passing back the entire set of tests including those that did not fail.

ChuckJonas replied:
Why?

This method doesn't need to return anything, just modify the elements have failed and need the additional information.

fishygeek91 replied:
So after I modify should I insert again? As of now I am calling this method in "processTestResults" and inserting the returned sObjects

ChuckJonas replied:
you don't need to run DML again.  The basic steps are:

- create the records
- find any failures
- determine lastSuccess/firstFailure
- set fields
- insert records

Your code is doing this, but just looping over records and making additional assignments needlessly

ChuckJonas commented on force-app/main/default/objects/Test_Run_Method_Result__c/fields/Failure_Audit__c.field-meta.xml:
```diff
@@ -0,0 +1,16 @@
+    <description>Displays SOQL query for failed test audit</description>
+    <externalId>false</externalId>
+    <formula>IF(Method_Pass__c == false,
+&quot;[SELECT Id, Action, CreatedBy.Name, CreatedDate, Display, Section FROM SetupAuditTrail WHERE CreatedDate &gt; &quot; &amp; TEXT(Last_Success__c) &amp; &quot; AND CreatedDate &lt; &quot; &amp; TEXT(First_Failure__c) &amp; &quot;]&quot;,
```
Look and see if there are any other fields on the setup audit trail that might be helpful to display

ChuckJonas reviewed (changes requested):
@fishygeek91 looks good, but there are a handful of changes that need to be made to optimize the code.

LMK if you have any questions

fishygeek91 committed:
-added Map structure to optimize inner for loop

ChuckJonas commented on force-app/main/default/classes/TestRunProcessor.cls:
```diff
@@ -108,6 +108,46 @@ public with sharing class TestRunProcessor implements Schedulable {
+            if (failedTests.containsKey(f.Method_Name__c)) {
+                List<Test_Run_Method_Result__c> tempList = failedTests.get(f.Method_Name__c);
+                tempList.add(f);
+                failedTests.put(f.Method_Name__c, tempList);
```
you don't need to `put` the list back in.  The map already contains this exact reference

ChuckJonas commented on force-app/main/default/classes/TestRunProcessor.cls:
```diff
@@ -108,6 +108,46 @@ public with sharing class TestRunProcessor implements Schedulable {
         }
 
-        insert results;
+        List<Test_Run_Method_Result__c> result = processPassFailDates(results);
```
result === results

This assignment effectively does nothing as `processPassFailDates` returns the same list that was passed in

fishygeek91 committed:
-edited TestRunProcessor to get rid of redundant references
-changed unit test around a little(more to come)

fishygeek91 committed:
-updated TestRunProcessorTest to include more code coverage

fishygeek91 committed:
-removed unnecessary SOQL query
-made changes so helper methods are more helpful
-made some comments
-changed some variable names for easier reading

veritascs commented on force-app/main/default/classes/TestRunProcessor.cls:
```diff
@@ -107,7 +95,44 @@ public with sharing class TestRunProcessor implements Schedulable {
+            ORDER BY CreatedDate DESC
+        ];
+
+        for (Test_Run_Method_Result__c f : previousResults) {
+            if (previousResultsMap.containsKey(f.Method_Name__c)) {
+                List<Test_Run_Method_Result__c> tempList = previousResultsMap.get(f.Method_Name__c);
+                tempList.add(f);
+            }
+        }
```
unused

ChuckJonas replied:
@veritascs which line? This block  is populating the previous results map

veritascs replied:
Word, not unused! My bad.

fishygeek91 committed:
-changed First_Failure__c assignment when instance IS the first failure
